compiler/shared_ast/typing.ml

```diff
@@ -224,54 +211,121 @@ let record_type_error _ctx (A.AnyExpr e) t1 t2 =
+  match Mark.remove t1, Mark.remove t2 with
+  | TLit tl1, TLit tl2 ->
+    if tl1 <> tl2 then record_type_error ();
+    t2
+  | TArrow (targs1, tret1), TArrow (targs2, tret2) ->
+    let tret = unify tret1 tret2 in
+    let targs =
+      try List.map2 unify targs1 targs2
+      with Invalid_argument _ ->
+        record_type_error ();
+        targs2
+    in
+    TArrow (targs, tret), pos2
+  | TTuple ts1, TTuple ts2 ->
+    let ts =
+      try List.map2 unify ts1 ts2
+      with Invalid_argument _ ->
+        record_type_error ();
+        ts2
+    in
+    TTuple ts, pos2
+  | TStruct s1, TStruct s2 ->
+    if not (StructName.equal s1 s2) then record_type_error ();
+    t2
+  | TEnum e1, TEnum e2 ->
+    if not (EnumName.equal e1 e2) then record_type_error ();
+    t2
+  | TOption t1', TOption t2' -> TOption (unify t1' t2'), pos2
+  | TArray t1', TArray t2' -> TArray (unify t1' t2'), pos2
+  | TDefault t1', TDefault t2' -> TDefault (unify t1' t2'), pos2
+  | TAny t1b, TAny t2b ->
+    let _, t1, t2 = Bindlib.unmbind2 t1b t2b in
+    unify t1 t2
+  | TAny t1b, _ ->
+    let _, t1 = Bindlib.unmbind t1b in
+    unify t1 t2
+  | _, TAny t2b ->
+    let _, t2 = Bindlib.unmbind t2b in
+    unify t1 t2
+  | TVar v1, TVar v2 -> (
+    if Bindlib.eq_vars v1 v2 then t2
+    else
+      match Env.get_tvar env v1, Env.get_tvar env v2 with
+      | None, None ->
+        Env.set_tvar env v1 t2;
+        t2
+      | Some (TVar v3, _), Some ((TVar v4, _) as t2) when Type.Var.equal v3 v4
+        ->
+        t2
+      | Some (TVar v3, _), None when Type.Var.equal v2 v3 -> t2
+      | None, Some (TVar v3, _) when Type.Var.equal v1 v3 -> t1
+      | Some t1, Some t2 ->
+        let t = unify t1 t2 in
+        Env.set_tvar env v1 t;
+        Env.set_tvar env v2 t;
+        t
+      | Some t1, None ->
+        if Type.Var.Set.mem v2 (Type.free_vars t1) then
+          Message.error ~internal:true ~pos:(Expr.pos e)
+            "Recursive type detected: %a(%a) = %a" Type.Var.format v1
+            Type.format t1 Type.format t2
+        else (
+          Env.set_tvar env v2 t1;
+          t1)
+      | None, Some t2 ->
+        if Type.Var.Set.mem v1 (Type.free_vars t2) then
+          Message.error ~internal:true ~pos:(Expr.pos e)
+            "Recursive type detected: %a(%a) = %a" Type.Var.format v2
+            Type.format t2 Type.format t1
+        else (
+          Env.set_tvar env v1 t2;
+          t2))
+  | TVar v1, _ ->
+    let t =
+      match Env.get_tvar env v1 with None -> t2 | Some t1 -> unify t1 t2
+    in
+    Env.set_tvar env v1 t;
+    t
+  | _, TVar v2 ->
+    let t =
+      match Env.get_tvar env v2 with None -> t1 | Some t2 -> unify t1 t2
+    in
+    Env.set_tvar env v2 t;
+    t
+  | TClosureEnv, TClosureEnv -> t2
+  | ( ( TLit _ | TArrow _ | TTuple _ | TStruct _ | TEnum _ | TOption _
+      | TArray _ | TDefault _ | TClosureEnv ),
+      _ ) ->
+    record_type_error ();
+    t2
 
-let lit_type (lit : A.lit) : naked_typ =
+let unify'
```

Comment:
Shouldn't this be `unify` and the other `unify'` or better `unify_and_return_repr`?